Evaluation (#122)
NVIDIA-AI-Blueprints/vulnerability-analysis: README.md

Additional output options will be added in the future.

## Evaluation
In this example evaluation workflow, we evaluate the pipeline's ability to identify vulnerability **status** and **label**.

Each CVE has one ground truth status and one corresponding label.
Valid statuses include: `AFFECTED` or `NOT AFFECTED`.
Valid labels include: `vulnerable` if the CVE status is `AFFECTED`, or one of 10 VEX statuses if the status is `NOT AFFECTED` (see **Justification Assignment** under the [Key Components section](#key-components)).

To run evaluaton, run `nat eval` using the `config-eval.yml` configuration.

```
nat eval --config_file=configs/config-eval.yml
```

This will write the files `status_accuracy_output.json` and `label_accuracy_output.json` to the configured output directory (`.tmp/evaluators` in this example), containing accuracy scores for status and label respectively. Scores range from 0 to 1, with 0 indicating no CVEs were categorized correctly, and 1 indicating all CVEs were categorized correctly.

> Under the hood, the source code of the evaluator being run can be found at `eval/evaluators/accuracy.py`; this custom evaluator is registered in the NeMo Agent Toolkit as an evaluator of type `_accuracy`.  In `config-eval.yml`, under the `evaluators` section, we have defined two evaluations to be run: `status_accuracy` and `label_accuracy`. However, the underlying code for these evaluations is the same. In other words, they use the same evaluator, which accepts either `status` or `label` as a configurable field input.

Reviewer: by `_accuracy` do we mean `accuracy`?

     ...
```

In `config-eval.yml`, configure the `file_path` field to specify the dataset to use. You can also configure the evaluation output directory using the `output_dir` variable.

An example evaluation experiment has been set up to run `eval_dataset` which runs test sets of CVEs on the morpheus_23.11-runtime and morpheus_24.03-runtime images.

```yml
file_path: data/eval_datasets/eval_dataset.json
```

### Running Multiple Evaluations

NeMo Agent Toolkit provides a `--reps` flag, which can be used to conveniently run an evaluation item multiple times. Note that **caching must be disabled first** (see [Disable Caching](#disable-caching) below) in order for evaluation repetitions to yield differing results.

```
nat eval --config_file=configs/config-eval.yml --reps=3
```

#### Visualizing Score Distribution Across Multiple Runs

> NeMo Agent Toolkit enables flexibility in defining evaluation output format, as well as flexibility in consuming outputs (eg. uploading to observability platforms like Phoenix, or running a custom visualization script). In this example, we use a custom script in `eval/visualizations/box_and_whisker_plot.py` to plot the distribution of accuracy outputs after running multiple repetitions. The script reads from the evaluation results written at your configured output directory. It can be run like this:

Reviewer: ```suggestion
NeMo Agent Toolkit enables flexibility in defining the evaluation output format, as well as flexibility in consuming outputs (e.g., uploading to observability platforms like Phoenix, or running a custom visualization script). In this example, we use a custom script in `eval/visualizations/box_and_whisker_plot.py` to plot the distribution of accuracy outputs after running multiple repetitions. The script reads from the evaluation results written at your configured output directory. It can be run like this:
```

You can also implement custom TTC strategies other than majority voting by following the NeMo Agent Toolkit [Test Time Compute reference](https://docs.nvidia.com/nemo/agent-toolkit/latest/reference/test-time-compute.html).

## Evaluation

> Evaluation is an essential part of developing agentic workflows. In this section, we use NeMo Agent Toolkit's Evaluation feature to implemenent an example evaluation pipeline that demonstrates best practices for performing evaluation on the vulnerability analysis workflow. There are two custom evaluators included in this example, which provide an extensible template for developers to add their own custom evaluators that target their desired metrics.

Reviewer 2: Minor wording suggestion:
```suggestion
Evaluation is an essential part of developing agentic workflows. In this section, we use NeMo Agent Toolkit's Evaluation feature to implemenent an example evaluation pipeline that demonstrates best practices for performing evaluation on the vulnerability analysis workflow. There are two example custom evaluators included in this blueprint, which provide an extensible template for developers to add their own custom evaluators that target their desired metrics.
```

To run evaluation, run `nat eval` using the `config-eval.yml` configuration.

```
nat eval --config_file=configs/config-eval.yml
```

This will run the two custom evaluators, which measure accuracy and consistency. Evaluation results can be viewed in `.tmp/evaluators`.

Visit the [Example Custom Evaluators](#example-custom-evaluators) section to learn more details about these evaluators.

> To extend this evaluation example, developers should also write their own custom evaluators to calculate desired metrics. Read more at the NeMo Agent Toolkit documentation for [custom evaluators](https://docs.nvidia.com/nemo/agent-toolkit/1.2/reference/evaluate.html#adding-custom-evaluators). Additionally, developers should provide their own test sets, which contain the ground truth data that the pipeline's output is compared against. See [Inputting Evaluation Test Sets](#inputting-evaluation-test-sets) for details.

Reviewer 2: Standardize terminology from "test set" to "evaluation dataset"
```suggestion
To extend this evaluation example, developers should also write their own custom evaluators to calculate desired metrics. Read more at the NeMo Agent Toolkit documentation for [custom evaluators](https://docs.nvidia.com/nemo/agent-toolkit/1.2/reference/evaluate.html#adding-custom-evaluators). Additionally, developers should provide their own evaluation datasets, which contain the ground truth data that the pipeline's output is compared against. See [Inputting Evaluation Datasets](#inputting-evaluation-datasets) for details.
```

> 

Reviewer 2: I highly recommend explicitly calling out that these examples are not a part of the stable API and might change in future releases. This helps set expectations in case our eval process evolves as we start using it ourselves :)

```suggestion

> [!NOTE]
> The custom evaluators and evaluation dataset format shown below are example implementations and not part of the stable API. Teams should use these as a starting point and adapt them for their specific evaluation needs. These examples may be updated in future releases.
```

### Inputting Evaluation Test Sets
Evaluation test sets containing ground truths should be put in the `data/eval_datasets/` directory. This file path is configurable via the `eval.general.dataset.filepath` field. 

Each test set is a JSON file which represents an evaluation experiment. Experiments can be run across multiple containers. An example test set `eval_dataset.json` is provided, which includes two containers (morpheus_23.11 and morpheus_24.03) which have a few CVEs each. The structure of the test set is as follows:

```
dataset_id
dataset_description
containers:
     containerA
          metadata
          ground-truth: [list of CVEs]
     containerB
          metadata
          ground-truth: [list of CVEs]
     ...
```

> Developers can input their own test sets by following this file structure.

Reviewer 2: Moved detailed dataset description from above to below
```suggestion
Evaluation datasets can contain ground truth data across multiple containers. Each entry in the `containers` field consists of input info for a container, along with a list of CVEs and their corresponding ground truth labels. An example test set `eval_dataset.json` is provided, which includes two containers (`morpheus:23.11-runtime` and `morpheus:24.03-runtime`) which have a few CVEs each. Developers can input their own test sets by following this file structure.
```

> A custom parsing script (`eval/parse_eval_input.py`)is used to transform test sets into a format ingestible by the workflow. The path to this script is configured in `eval.general.dataset.function`. Advanced developers can choose to modify the test set input structure by creating their own parsing script, and then updating the configuration field with their own script.

Reviewer 2: Minor missed space, and adding link to NAT custom datasets docs
```suggestion
A custom parsing script (`eval/parse_eval_input.py`) is used to transform test sets into a format ingestible by the workflow. The path to this script is configured in `eval.general.dataset.function`. Advanced developers can choose to modify the test set input structure by creating their own parsing script, and then updating the configuration field with their own script. See the NAT documentation on [Custom Dataset Format](https://docs.nvidia.com/nemo/agent-toolkit/latest/reference/evaluate.html#custom-dataset-format) to learn more about customizing datasets for your evaluation needs.
```

Re-enable caching by setting the `NVIDIA_API_BASE` variable back to NGINX.
```
export NVIDIA_API_BASE=http://nginx-cache/nim_llm/v1
```

##### Option 2: Disable Caching on Specific Sections
This option enables developers to disable caching for specific parts of the workflow. It is useful for disabling caching on all model calls except the embedding model, in order to avoid the time-consuming embedding process. This approach is also useful for in-depth evaluation experiments; for example, a developer may want to disable caching for just the `justify_llm` in order to test consistency of the justification part of the workflow without re-running everything before that point. 

In `config-eval.yml`, assign the `base_url` field of the target section to `"https://integrate.api.nvidia.com/v1"`. We recommend creating multiple config files to keep track of different evaluation experiments. To re-enable caching, simply assign the `base_url` back to `${NVIDIA_API_BASE:-https://integrate.api.nvidia.com/v1}`.

> For quick testing, developers can also change configuration in the CLI using the `--override` flag, which uses dot notation. This change will not persist across future runs. Below is an example that disables caching on the checklist_llm.

Reviewer 2: Removing the note about disabling caching for embedding model, + minor wording/formatting.

```suggestion
##### Option 2: Disable Caching on Specific Components
This option enables developers to disable caching for specific parts of the workflow. This approach is useful for in-depth evaluation experiments; for example, a developer may want to disable caching for just the `justify_llm` in order to test consistency of the justification part of the workflow without re-running everything before that point.

In `config-eval.yml`, assign the `base_url` field of the target components to `"https://integrate.api.nvidia.com/v1"`. We recommend creating multiple config files to keep track of different evaluation experiments. To re-enable caching, simply assign the `base_url` back to `${NVIDIA_API_BASE:-https://integrate.api.nvidia.com/v1}`.

For quick testing, developers can also change configuration in the CLI using the `--override` flag, which uses dot notation. This change will not persist across future runs. Below is an example that disables caching on the `checklist_llm`.
```

```
nat eval --config_file=configs/config-eval.yml --reps=3 --override llms.checklist_llm.base_url https://integrate.api.nvidia.com/v1
```

### Example Custom Evaluators
In this section, the implementation of the two custom evaluators for accuracy and consistency are discussed in detail.

#### Simple Accuracy Evaluator
For the vulnerability analysis workflow, the most simple evaluation we can begin with is evaluating the accuracy of the `status` and `label` fields which the workflow outputs per CVE.

Each CVE has a ground truth `status`, as well as a `label` which corresponds to the status.
Valid statuses include: `AFFECTED` or `NOT AFFECTED`.
Valid labels include: `vulnerable` if the CVE status is `AFFECTED`, or one of 10 VEX statuses if the status is `NOT AFFECTED` (see **Justification Assignment** under the [Key Components section](#key-components)).

> For each CVE, if the workflow output matches the ground truth, a score of 1 is assigned. Otherwise, a score of 0 is assigned. The evaluator then takes the mean of all CVEs' scores to produce an average accuracy.

Reviewer 2: Minor: use "accuracy score" to refer to mean over CVEs, and avoid confusion with average accuracy over reps.
```suggestion
For each CVE, if the workflow output matches the ground truth, a score of 1 is assigned. Otherwise, a score of 0 is assigned. The evaluator then takes the mean of all CVEs' scores to produce an accuracy score.
```

The evaluator also takes in a `duplicates_policy` field. This is to account for instances where, if an input set is an aggregation of ground truths from different sources, the set may contain duplicate CVEs with conflicting answers. For example, a CVE may be listed twice in the dataset as both `AFFECTED` and `NOT AFFECTED`, with different label justifications. We provide three policies for handling duplicates:
- `drop_all`: Any CVE with conflicting ground truth will be omitted and not contribute towards evaluation results.
- `keep_first`: Use the first encountered instance of a CVE as the truth, disregard all other instances.
- `keep_all`: Any answer listed in the ground truth for a CVE is considered valid (eg. if the ground truth has two listings of a CVE, with one being `AFFECTED` and the other being `NOT AFFECTED`, either answer given by the pipeline will be considered correct).

##### Output Format
The final scores are written to the files `status_accuracy_output.json` and `label_accuracy_output.json`. The output will include an `average_score` overview field, and a an `eval_output_items` field which gives a more detailed breakdown.

###### average_score Field

> The `average_score` field contains a dict with metrics corresponding to `pandas.Dataframe.describe` output. The goal of this field is to capture the consistency of runs, when a user uses the NeMo Agent Toolkit `--reps` option to perform multiple run of the experiment. If only one run is performed, `count` and `mean` would be the only meaningful metrics; for example, if the run yielded accuracy of 0.8, the `average_score` field would look something like this:

Reviewer 2: Wording suggestions for clarity
```suggestion
The `average_score` field contains a dict with metrics corresponding to `pandas.Dataframe.describe` output. The goal of this field is to aggregate the accuracy score across all runs, e.g., when using the `--reps` option. If only one run is performed, `count` and `mean` would be the only meaningful metrics; for example, if the run yielded accuracy of 0.8, the `average_score` field would look like this:
```

```json
"average_score": {
    "count": 1,
    "mean": 0.8,

>     "std": NaN,

Reviewer 2: I think we output null now?
```suggestion
    "std": null,
```

    "min": 0.8,
    "25%": 0.8,
    "50%": 0.8,
    "75%": 0.8,
    "max": 0.8
  }
```

> Meanwhile, if there were multiple runs, the `average_score` field would capture the distribution of runs. A script is provided in `eval/visualizations/box_and_whisker_plot.py` to visualize this distribution as a box and whisker plot. It is especially helpful for comparing the plots of multiple experiments side-by-side, or comparing `status` accuracy against `label` accuracy. It can be run like so:

Reviewer 2: Do you have an example of a multi-output run distribution handy that we can include here to show the difference?

> Meanwhile, if there were multiple runs, the `average_score` field would capture the distribution of runs. A script is provided in `eval/visualizations/box_and_whisker_plot.py` to visualize this distribution as a box and whisker plot. It is especially helpful for comparing the plots of multiple experiments side-by-side, or comparing `status` accuracy against `label` accuracy. It can be run like so:

Reviewer 2: Minor wording suggestion.

Also, do you think it would be helpful include an example multi-run output to illustrate the difference?
```suggestion
Meanwhile, if there were multiple runs, the `average_score` field would capture the accuracy distribution across runs. A script is provided in `eval/visualizations/box_and_whisker_plot.py` to visualize this distribution as a box and whisker plot. It is especially helpful for comparing the plots of multiple experiments side-by-side, or comparing `status` accuracy against `label` accuracy. It can be run like so:
```

```
python3 src/vuln_analysis/eval/visualizations/box_and_whisker_plot.py \
  .tmp/evaluators/accuracy-output-1.json \
  .tmp/evaluators/accuracy-output-2.json \
  --title "Accuracy: Experiment 1 vs. Experiment 2" \
  --save src/vuln_analysis/eval/visualizations/my-plot.png \
  --labels my-experiment1 my-experiment2
```
Users can list any number of output files, as well as configure chart title, save location, and X-axis labels.

###### eval_output_items Field
The `eval_output_items` field gives a more detailed output of the experiment. It lists the runs, giving each run a unique `id`, a `score` (the accuracy score for that run), and a `reasoning` field.

> This last field breaks the run down by container, and lists the specific CVEs that were run in each container (`question`), the ground truth answers `[target field]_answer`, and the answers generated by the workflow (`generated_[target field_answer`).

Reviewer 2: Minor: missing ]
```suggestion
This last field breaks the run down by container, and lists the specific CVEs that were run in each container (`question`), the ground truth answers `[target field]_answer`, and the answers generated by the workflow (`generated_[target field]_answer`).
```

#### Consistency Evaluator
This evaluator measures the consistency of results across repeated runs. It can be configured to measure consistency of the `status` or `label` fields.

Consistency is defined as: `∑p²` per CVE, where *p* is the probability of each unique status or label. For example, if the status for a CVE across 3 runs are `(AFFECTED, NOT_AFFECTED, NOT_AFFECTED)`, then the probabilities become `p(AFFECTEd, NOT_AFFECTED) = {1/3, 2/3}`. Therefore, the consistency score is `(1/3)² + (2/3)² = 0.56`. A score of 1.0 means perfect consistency (all runs yielded the exact same values). The evaluator outputs a consistency score for each CVE.

**Note: This evaluator should only be used if evaluation is being run with multiple repetitions. An error will occur if this criteria is not met.** To enable this evaluator, uncomment the last section in `config-eval.yml`.

##### Configuration Options

> The accuracy evaluator is implemented in `eval/evaluators/consistency.py` and registered in NeMo Agent Toolkit with the unique identifer (`_type`) of "`consistency`".

Reviewer 2: Some final minor formatting suggestions
```suggestion
Consistency is defined as: `∑p²` per CVE, where *p* is the probability of each unique status or label. For example, if the status for a CVE across 3 runs are `(AFFECTED, NOT_AFFECTED, NOT_AFFECTED)`, then the probabilities become `p(AFFECTED) = 1/3`, `p(NOT_AFFECTED) = 2/3`. Therefore, the consistency score is `(1/3)² + (2/3)² = 0.56`. A score of 1.0 means perfect consistency (all runs yielded the exact same values). The evaluator outputs a consistency score for each CVE.

> [!NOTE]
> This evaluator should only be used if evaluation is being run with multiple repetitions. An error will occur if this criteria is not met. To enable this evaluator, uncomment the last section in `config-eval.yml`.

##### Configuration Options
The accuracy evaluator is implemented in `eval/evaluators/consistency.py` and registered in NeMo Agent Toolkit with the unique identifer (`_type`) of `consistency`.
```

If you run out of credits for the NVIDIA API Catalog, you will need to obtain more credits to continue using the API. Please contact your NVIDIA representative to get more credits added.

### Mac Workarounds
Development on macOS is supported on a best-effort basis and may require workarounds. One current known issue is that running the project docker container causes a `ModuleNotFoundError: No module named 'numpy.distutils'` related to faiss. To work around this, in the `Dockerfile` change the `ARG PYTHON_VERSION=3.12` line to `ARG PYTHON_VERSION=3.11`. The full issue is tracked [here](https://github.com/NVIDIA-AI-Blueprints/vulnerability-analysis/issues/115).

> 

Author: Done. Will link new PR here after I finish going through eval PR
